### ISSUE TITLE
Is protobuf-es compatible with Deno?

### ISSUE THREAD

**Issue by taras** [bufbuild/protobuf-es#293]
I came across [protobuf-es announcement](https://buf.build/blog/protobuf-es-the-protocol-buffers-typescript-javascript-runtime-we-all-deserve) and it looks great. The analysis was excellent as well. Since you have native ES support, have you tried it Deno? Thank you, keep up the great work.

**Comment by taras** [comment]
> we are all very excited about Deno. Generally, our runtime should work fine with Deno because we only use a few standard APIs. 

Sweet!

> I'd be also very curious how you run a plugin with Deno. Would you need an equivalent to [this script](https://github.com/bufbuild/protobuf-es/blob/main/packages/protoc-gen-es/bin/protoc-gen-es) specific to Deno, or are there ways to invoke an npm package "bin"?

@cowboyd how do you think this should work based on what we've been doing with Deno and lessons learned from GraphQL?

**Comment by cowboyd** [comment]
@timostamm On point (3), it might be a non-issue? That's because importing plain JavaScript sources in Deno "just works." In fact, the entire reason Deno requires the file extension is so that it know whether to invoke the TypeScript compiler or not.

As you said, (1) and (2) can be worked around using import maps, but most `npm` modules are available on one of the "un-packagers" such as https://esm.sh or https://skypack.dev

Our approach is to write the source code in Deno itself because its such a great developer experience (it is very similar to go-lang) and then build the Npm package as a downstream artifact.

**Comment by timostamm** [comment]
@cowboyd, I think I never tried it with generated `.js` and `.d.ts` files 🤔 

Yup, it works fine, just with an import map. I can make a request to our demo service with a [connect-web](https://github.com/bufbuild/connect-web) client. 
See [this gist](https://gist.github.com/timostamm/6f4200353052fd4a95204a3d6863e4b3).

Will this still properly lint (including typical IDE setups)? Do you think there are any other problematic areas?

**Comment by NfNitLoop** [comment]
This seems to be working well for me.  I've hacked together a Deno task [here] that:

 * uses an embedded npm package(.json) to install `protoc-gen-es` 
 * calls `protoc` w/ options to use `protoc-gen-es`
 * fixes up the import path in the result to use a module from esm.sh.

[here]: https://github.com/NfNitLoop/feoblog-client/commit/b07e6fb75a1351611bd22a36404463472900108e#diff-d64dc538a8095423a956716c63fbb0a89fd0e792069572caad782b26e7e61ea0

The result gives code that I'm able to use directly in Deno. 🎉   (And later I'm going to see if this allows me to also use [dnt] to also publish this client as an npm package.) 

[dnt]: https://deno.land/manual@v1.26.0/node/dnt

Minor feature requests along the way:

* A way to tell the protoc-gen-es plugin to import from somewhere else.  
  I work around it by just doing a replaceAll on the result.
* A way to install and run `protoc-gen-es` with Deno instead of node.  
  While https://esm.sh/@bufbuild/protobuf@0.3.0 worked, https://esm.sh/@bufbuild/protoc-gen-es@0.3.0 does not. Maybe it's using some node-specific features that esm.sh doesn't know how to shim?
* Access helpers for `oneof` fields.  
   (Should probably make this a separate issue, just including here for context)
   In migrating from <https://github.com/thesayyn/protoc-gen-ts>, it seems odd that I can't check for a `.post` directly on my object, but have to do a [type assertion dance][1] with `case` and `value`.  

[1]: https://github.com/NfNitLoop/feoblog-client/commit/b07e6fb75a1351611bd22a36404463472900108e#diff-e4ad9b6995ee7a884aa6f86ffb808e894cd5e288fac02dc0a21537abdf6b5ed2R19-R22

Instead of 

```typescript
  itemType: {
    value: Post;
    case: "post";
  } | {
    value: Profile;
    case: "profile";
  } | {
    value: Comment;
    case: "comment";
  } | { case: undefined; value?: undefined } = { case: undefined };
```

Wouldn't you still get the same type safety with this?

```typescript
 itemType: {
    post: Post;
  } | {
    profile: Profile;
  } | {
    comment: Comment;
  } | {} = {};
```

**Comment by NfNitLoop** [comment]
Ah yes, one more oddity to point out:

[Here], I discovered that the code tries to access an environment variable named `BUF_BIGINT_DISABLE`. I expect this is a thing in Node. In Deno, attempting to read an environment variable stops the application to prompt the user for permissions.  

[here]: https://github.com/NfNitLoop/feoblog-client/commit/b07e6fb75a1351611bd22a36404463472900108e#diff-81342196266f9373572c759839aa5a9ab9942af52ce56316b4855b3df8db588aR3-R5

**Comment by unicomp21** [comment]
Crossing my fingers, would be awesome if this works!

**Comment by jcready** [comment]
> Wouldn't you still get the same type safety with this?
> ```ts
>  itemType: {
>     post: Post;
>   } | {
>     profile: Profile;
>   } | {
>     comment: Comment;
>   } | {} = {};
> ```

No.

```ts
type exampleOneof = {
  foo: string;
} | {
  bar: number;
} | {
  baz: boolean;
} | {};

declare var example: exampleOneof;

// Can set excess properties
example = { excess: 'property' }; // no error

// Can set INCORRECT values 😱
example = { foo: 1 }; // no error

// Can set MULTIPLE fields 😱
example = { foo: 'both are set', bar: 1 }; // no error

// Can't check which is set using property access
if (example.foo) {}
//           ^ Property 'foo' does not exist on type 'exampleOneof'.
//             Property 'foo' does not exist on type '{}'.(2339)
if (typeof example.foo !== 'undefined') {}
//                  ^ Property 'foo' does not exist on type 'exampleOneof'.
//                    Property 'foo' does not exist on type '{}'.(2339)

// Using `hasOwnProperty()` doesn't refine
if (example.hasOwnProperty('foo')) { example.foo }
//                                            ^ Property 'foo' does not exist on type 'exampleOneof'.
//                                              Property 'foo' does not exist on type '{}'.(2339)

```

**Comment by NfNitLoop** [comment]
For those interested, I got this working "in Deno", here: 
https://github.com/NfNitLoop/feoblog-client/tree/aeb18d564909e99f37b3a14d2090abbdc2f1c322/private/protobuf

The "in Deno" is in quotes because I'm currently temporarily using npm to install protobuf-es into a local directory, then use that with `protoc` to convert my `.proto` file into `.ts`.  Then it works fine from Deno, and also survived the translation by [DNT] back into an npm package.  This lets my Deno package be my canonical source, but I get to use it from npm too when I have to. :p 

[DNT]: https://github.com/denoland/dnt#readme

--- 

@jcready -- thanks for the example. I forgot how lax `{}` is in TypeScript. I've got a better suggestion over in #337.  :)

**Comment by smaye81** [comment]
Adding here for posterity also.  When Deno support is added, we should support allowing `rewrite_imports` to rewrite to URLs.

See https://github.com/bufbuild/protobuf-es/issues/483 for the original issue requesting this.

**Comment by timostamm** [comment]
Deno v2 supports importing from npm packages with the ["npm:" specifier](https://docs.deno.com/examples/npm/). We're updating the plugin option `rewrite_imports` to support the prefix in https://github.com/bufbuild/protobuf-es/pull/998 (see https://github.com/bufbuild/protobuf-es/issues/992 for an example). 

Deno expects the extension `.ts` for file imports, which is supported with the plugin option `import_extension=ts`. Going forward, it may be worth it to support `target=deno` to automatically set `target=ts`, `import_extension=ts`, and `rewrite_imports` to add the npm specifier and a version for `@bufbuild/protobuf`.

**Comment by yordis** [comment]
@timostamm this sounds that could be, most likely, close now

**Comment by timostamm** [comment]
I just gave Deno 2.2.6 a try with a basic example, starting out with `deno init my_project`:

```
.
└── my_project
    ├── buf.gen.yaml
    ├── deno.json
    ├── deno.lock
    ├── gen
    │   └── example_pb.ts
    ├── main.ts
    ├── main_test.ts
    └── proto
        └── example.proto
```

The packages for the runtime library and plugin install without issues with `deno install npm:@bufbuild/protobuf`. Here's deno.json:

```
{
  "tasks": {
    "dev": "deno run --watch main.ts",
    "generate": "buf generate"
  },
  "imports": {
    "@bufbuild/protobuf": "npm:@bufbuild/protobuf@^2.2.5",
    "@bufbuild/protoc-gen-es": "npm:@bufbuild/protoc-gen-es@^2.2.5",
    "@std/assert": "jsr:@std/assert@1"
  }
}
```

Deno does not make npm "binaries" available on the PATH in tasks. `deno task generate` fails with `Failure: plugin protoc-gen-es: exec: "protoc-gen-es": executable file not found in $PATH`.

This can be worked around by specifying a full command for the plugin in buf.gen.yaml:

```yaml
# Learn more: https://buf.build/docs/configuration/v2/buf-gen-yaml
version: v2
inputs:
  - directory: proto
clean: true
plugins:
  - local: ["deno", "run", "npm:@bufbuild/protoc-gen-es"]
    opt:
      - target=ts
      - import_extension=ts
      - rewrite_imports=@bufbuild/protobuf:npm:@bufbuild/protobuf
      - rewrite_imports=@bufbuild/protobuf/codegenv1:npm:@bufbuild/protobuf/codegenv1
    out: gen
    include_imports: true
```

Note that I'm also setting `import_extension=ts`, and rewrite imports to add the npm specifier.

When running generated code, deno requests env access to BUF_BIGINT_DISABLE. This can be worked around by permitting access with a flag: `deno run --allow-env=BUF_BIGINT_DISABLE main.ts`.


main.ts:

```ts
import { create } from "npm:@bufbuild/protobuf";
import { UserSchema } from "./gen/example_pb.ts";

export const user = create(UserSchema, {
  firstName: "John",
  lastName: "Doe",
  active: true,
});

// Learn more at https://docs.deno.com/runtime/manual/examples/module_metadata#concepts
if (import.meta.main) {
  console.log(user);
}
```

proto/example.proto is packages/protobuf-example/proto/example.proto from protobuf-es.

**Comment by timostamm** [comment]
> Is protobuf-es compatible with Deno?

Judging from https://github.com/bufbuild/protobuf-es/issues/293#issuecomment-2765625581 it appears to be compatible, but there clearly are some rough edges.

Before we can close this issue as completed, we should have some documentation and continuous (conformance) tests. 
Right now, we don't have the bandwidth to implement and maintain those additions, but maybe we do in the future.

So for the time being, I'd like to encourage everyone interested in Protobuf-ES on Deno to give it a try, and let us know of any other rough edges in this issue.

(Really love what Deno is doing with built-in test runners, linting, and tasks. It's a game-changer having these features built into the tool 😍 )
